### DISCUSSION TITLE
A question regarding the various "setMinMaxBitrate"

### DISCUSSION THREAD

**Discussion by neilyoung** [Kurento/kurento#17] [Community]
The upgrade statements says:

<img width="574" alt="image" src="https://user-images.githubusercontent.com/731020/227282256-4a8cce87-5061-48cc-9ac7-4c514b6abc2e.png">

Since my code is still compatible to 6.18 I'm reflecting this like so:

```
                if (this.versionCompare(this.kmsVersion, '6.18.0') < 0) {
                    await webRtcEndpoint.setOutputBitrate(config.maxBandwidth * 1000)
                }
                else {
                    await webRtcEndpoint.setMinEncoderBitrate(config.maxBandwidth * 1000 / 10)
                    await webRtcEndpoint.setMaxEncoderBitrate(config.maxBandwidth * 1000)
                }
```

Branch passed here the "else" branch.

I'm now having a room setting `config.maxBandwidth`, which defines the bandwidth to be configured at several occasions in kBps. It is set to 3000 currently.

In a publisher-filter-subscriber scenario this now leads to a publisher stream of constant 3MBit/s, which is then treated by the filter (transcoded before the delivery to OpenCV) and re-encoded, same standard H.264.

In previous version 6.18. I was able to see, that - if no movement is in the publisher video - the subscriber's input bitrate was about 300kBps, which matches the `setMinEncoderBitrate` setting. And it did follow up to 3MBit/s,. 

But regardless of how many changes I produce publisher side in KMS 7.0.0, these subscriber video bitrate never exceeds 1 MBit/s.

EDIT: And the video quality subscriber side is accordingly bad.
Why is that so?

**Comment by neilyoung** [comment]
If I config it in /etc/kurento/modules/kurento/MediaElement.conf.ini - perfect. 3M constant.

But this is a room setting, I would like to set it programmatically.

**Comment by neilyoung** [comment]
Confirmed. The programmatic setting of min/max encoder bitrate does not work :( 

I swear, it worked in the compiled version I made last week from 7.0.0-dev. I was using `setMin/MaxOutputBitrate` that time, but these APIs are gone now.

EDIT: That being said I see a regression here.

EDIT2: I also double checked that I'm setting the proper values:

2023-03-23T17:32:48.772Z info: setEncoderBitrate 3000000 setMinEncoderBitrate 300000 setMaxEncoderBitrate 3000000

**Comment by j1elo** [comment]
This does look like a regression that I'll have to look into. The `OutputBitrate` properties had the issue that the average bitrate of the encoder was hardcoded to 300 kbps, and that gave weird results especially for people who used the API with a very low minimum, a very high maximum, and never saw the average bitrate grow above 300 Kbps (because, well, it was hardcoded that way).

So the `EncoderBitrate` API was added to be a replacement that works as expected. Namely:

If you set the minimum at e.g. 100 Kbps, the maximum at 10,000 Kbps, and the average target at 5,000 Kbps, you should see an average bitrate output of 5 Mbps which depending on the scene, might raise up to 10 Mbps or go as low as 100 Kbps (always according to the decisions made by the encoder).

If that's not the behavior you're seeing, then something is definitely wrong in there. Please try that instead of setting everything to the same value.

Another thing that might be happening is that the min/max collision detection has a bug and when setting everything to the same value, the code is becoming a bit crazy... I'll look at the logic behind this to see if there is anything suspicious.

**Comment by neilyoung** [comment]
Do I have to set these values as "kBps" or as "bps"?

**Reply by j1elo** [reply]
As "bps" - see JS docs at [BaseRtpEndpoint.getEncoderBitrate](https://doc-kurento.readthedocs.io/en/latest/_static/client-jsdoc/module-core_abstracts.BaseRtpEndpoint.html#getEncoderBitrate)

**Reply by neilyoung** [reply]
Yepp. But it doesn't change much. I can re-read my settings, so they arrive somewhere. But I can't see an effect.

Any hint, what I could trace in KMS in order to see more?

**Comment by neilyoung** [comment]
I tried "kbps" first.

Publisher side:

<img width="356" alt="image" src="https://user-images.githubusercontent.com/731020/227330662-32401ccf-318f-46fb-82f0-3191bb2e3be1.png">

Subscriber side:

<img width="332" alt="image" src="https://user-images.githubusercontent.com/731020/227330885-7194f0ac-6396-474c-9fa9-6267fe7df9ab.png">

Setting:

2023-03-23T19:33:25.459Z info: setEncoderBitrate 5000 setMinEncoderBitrate 100 setMaxEncoderBitrate 10000

Verdict: Doesn't work

Will try "bps" next.

**Comment by neilyoung** [comment]
With "bps"

Publisher:

<img width="342" alt="image" src="https://user-images.githubusercontent.com/731020/227332690-6b2c921d-5095-40df-8f2d-57264852da11.png">


Subscriber:

<img width="350" alt="image" src="https://user-images.githubusercontent.com/731020/227332872-0c6dcbbb-3083-4bac-8f0c-dbff4ca336da.png">


Setting: 

2023-03-23T19:37:41.374Z info: setEncoderBitrate 5000000 setMinEncoderBitrate 100000 setMaxEncoderBitrate 10000000

Verdict: I don't know. Seems to not work

**Comment by neilyoung** [comment]
I was wrong. It does also not work with the kms7-dev version I installed last week. Very weird...

**Comment by neilyoung** [comment]
Q: I suppose the `kurento-client` 7.0.0 (npm) is not backward compatible to a 6.18.0 KMS installation, right? Especially because of the fact, that `setOutputBitrate` is no longer available?

**Comment by neilyoung** [comment]
Hi Juan, is there any progress here? Any hint, what to examine?

**Comment by neilyoung** [comment]
I really wonder, what the problem is: The publisher pumps with constant 3M, but the subscriber (transcoding) is not getting significantly more that 500K.

Even though I'm reportedly setting everything right:

2023-03-27T10:45:01.828Z info: getEncoderBitrate 3000000, getMinEncoderBitrate 1500000, getMaxEncoderBitrate 4500000

Why does this not work? It works fine, if I set the variables in the *.ini.

<img width="430" alt="image" src="https://user-images.githubusercontent.com/731020/227922136-78bddee7-694f-41d8-911f-68dc1ead4085.png">

<img width="398" alt="image" src="https://user-images.githubusercontent.com/731020/227922203-62665ce4-d544-4472-bb02-7ec055be83f1.png">

**Comment by neilyoung** [comment]
It is definitely only working with static config:

<img width="384" alt="image" src="https://user-images.githubusercontent.com/731020/227941492-a561d998-c156-4d4a-b865-8b9c8bd32f51.png">

For the API I'm now convinced, that there is a bug. For instance: The order of "setMin/Max" is deciding. If you set MIN before MAX you will get this warning:

2023-03-27T12:06:04,775792 25283 0x00007ffa93842700 warning kmselement                kmselement.c:1294 kms_element_set_property() <kmswebrtcendpoint1>  Ignoring requested bitrate min > max; setting max: 0

This is all so weird... Very disappointing....

**Comment by neilyoung** [comment]
Also a nice one:

```
2023-03-27T12:46:47,202074 33827 0x00007fbf7da5f700 warning kmselement                kmselement.c:1308 kms_element_set_property() <kmswebrtcendpoint4>  Ignoring requested bitrate max < min; setting min: 500000
2023-03-27T12:46:47,202933 33827 0x00007fbf7c25c700 warning kmselement                kmselement.c:1279 kms_element_set_property() <kmswebrtcendpoint4>  Ignoring out of bounds requested target bitrate; setting default: 300000
```

Or this:

```
2023-03-27T12:51:53,828216 36804 0x00007f55c9f46700 warning kmselement                kmselement.c:1279 kms_element_set_property() <kmswebrtcendpoint1>  Ignoring out of bounds requested target bitrate; setting default: 300000
```

on this sequence:

```
                    await webRtcEndpoint.setMaxEncoderBitrate(10000000)
                    await webRtcEndpoint.setMinEncoderBitrate(10000000)

                    await webRtcEndpoint.setEncoderBitrate(3000000)
```

Don't get it...

**Comment by neilyoung** [comment]
I think the only way out of this dilemma is to

a) drop all that more or less stupid validation
b) create a new setEncoderBitrates(min, max, target) API

**Comment by neilyoung** [comment]
OK, giving up here. This is entirely broken. Only fix configuration kind of works. "Kind of": It keeps the target bitrate, no min, no max. Any attempt to configure whatever bitrate later on (even with all verifications removed) does fail silently. Subscriber side this looks like so then (the received bitrate meanders around 300k, even all logs confirm the proper setup).

<img width="336" alt="image" src="https://user-images.githubusercontent.com/731020/228054454-daf3e0ed-e098-4eb8-987e-95bcff2047f4.png">

```
0:00:24.147464461 47087 0x7fe6b00022c0 DEBUG   KurentoMediaElementImpl MediaElementImpl.cpp:689:MediaElementImpl: Configured target video bitrate for media transcoding: 3000000 bps
0:00:24.147493000 47087 0x7fe6b00022c0 DEBUG   KurentoMediaElementImpl MediaElementImpl.cpp:694:MediaElementImpl: Configured minimum video bitrate for media transcoding: 500000 bps
0:00:24.147502721 47087 0x7fe6b00022c0 DEBUG   KurentoMediaElementImpl MediaElementImpl.cpp:699:MediaElementImpl: Configured maximum video bitrate for media transcoding: 4500000 bps
```

This is buggy 101

**Comment by j1elo** [comment]
Hi, I've been fixing other bug caused by newer versions of GStreamer: https://github.com/Kurento/kurento/commit/f226f46932c4dcc5f3379a7ef67e0bf24e5df089

Regarding this bug, please follow up in the bug tracker issue, https://github.com/Kurento/kurento/issues/18
